### PR TITLE
Replace liveness and readiness with initialization node controller

### PR DESCRIPTION
**Pull request by suket22**
**1. Issue, if available:**
https://github.com/aws/karpenter/issues/1135

**2. Description of changes:**
Earlier we would only delete worker nodes iff the kubelet would have a status of `NodeStatusNeverUpdated` for more than 15 minutes. If this were the case, it meant the Kubelet never connected to the API Server at all. 

We now delete worker nodes iff the kubelet is in NotReady status for more than 15 minutes **after startup**. The mechanism we use to discover that a node has only been started is by looking at a taint we apply during node object creation. Once the taint has been removed, we declare the startup as complete and no longer re-evaluate the node for deletion. We will re-introduce something like a liveness controller in the future if necessary but we want to be very careful with that to maintain static stability of the cluster.

**3. How was this change tested?**
Case 1 - I removed the CNI policy from the Node Role. Nodes were stuck in NotReady and I can see them being terminated after 15 minutes. 
```
2022-01-19T17:51:22.341Z	INFO	controller.provisioning	Batched 4 pods in 1.101064515s	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T17:51:22.551Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 4 pod(s) with instance type option(s) [c1.xlarge c4.2xlarge c3.2xlarge c6i.2xlarge c5ad.2xlarge c5d.2xlarge c5a.2xlarge c5.2xlarge c5n.2xlarge m3.2xlarge m5dn.2xlarge m5a.2xlarge t3.2xlarge m6i.2xlarge m5ad.2xlarge t3a.2xlarge m5n.2xlarge m4.2xlarge m5zn.2xlarge m5d.2xlarge]	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T17:51:24.811Z	INFO	controller.provisioning	Launched instance: i-01de791cd273012e1, hostname: ip-192-168-187-149.us-west-2.compute.internal, type: t3a.2xlarge, zone: us-west-2b, capacityType: on-demand	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T17:51:24.852Z	INFO	controller.provisioning	Bound 4 pod(s) to node ip-192-168-187-149.us-west-2.compute.internal	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T17:51:24.852Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T17:57:21.288Z	INFO	controller.provisioning	Batched 4 pods in 1.050060518s	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T17:57:21.650Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 4 pod(s) with instance type option(s) [c1.xlarge c4.2xlarge c3.2xlarge c5ad.2xlarge c5d.2xlarge c5a.2xlarge c5.2xlarge c6i.2xlarge c5n.2xlarge m3.2xlarge t3a.2xlarge m5zn.2xlarge t3.2xlarge m5.2xlarge m5a.2xlarge m5dn.2xlarge m6i.2xlarge m4.2xlarge m5ad.2xlarge m6a.2xlarge]	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T17:57:23.796Z	INFO	controller.provisioning	Launched instance: i-03c6fb55b68bb907f, hostname: ip-192-168-157-226.us-west-2.compute.internal, type: t3a.2xlarge, zone: us-west-2b, capacityType: on-demand	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T17:57:23.859Z	INFO	controller.provisioning	Bound 4 pod(s) to node ip-192-168-157-226.us-west-2.compute.internal	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T17:57:23.859Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T18:03:11.328Z	INFO	controller.provisioning	Batched 4 pods in 1.09442776s	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T18:03:11.719Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 4 pod(s) with instance type option(s) [c1.xlarge c3.2xlarge c4.2xlarge c5d.2xlarge c5ad.2xlarge c5.2xlarge c5a.2xlarge c6i.2xlarge c5n.2xlarge m3.2xlarge m5zn.2xlarge m5.2xlarge m5dn.2xlarge m6a.2xlarge m5n.2xlarge m5ad.2xlarge m4.2xlarge m6i.2xlarge m5a.2xlarge t3.2xlarge]	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T18:03:14.331Z	INFO	controller.provisioning	Launched instance: i-0b4949f10b9937bf9, hostname: ip-192-168-136-179.us-west-2.compute.internal, type: c5a.2xlarge, zone: us-west-2b, capacityType: on-demand	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T18:03:14.377Z	INFO	controller.provisioning	Bound 4 pod(s) to node ip-192-168-136-179.us-west-2.compute.internal	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T18:03:14.377Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "ee0d0b5", "provisioner": "default"}
2022-01-19T18:06:24.001Z	INFO	controller.node	Triggering termination for node that failed to transition to ready	{"commit": "ee0d0b5", "node": "ip-192-168-187-149.us-west-2.compute.internal"}
2022-01-19T18:06:24.035Z	INFO	controller.termination	Cordoned node	{"commit": "ee0d0b5", "node": "ip-192-168-187-149.us-west-2.compute.internal"}
2022-01-19T18:06:24.256Z	INFO	controller.termination	Deleted node	{"commit": "ee0d0b5", "node": "ip-192-168-187-149.us-west-2.compute.internal"}
```

Case 2 - I'm still looking to replicate the case, where a node does come up as healthy, but then goes into NotReady. I'm probably going to try a network partition to replicate this scenario but this is WIP. 

**4. Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [x] No

I don't think this impact our docs but I'm happy to add a callout somewhere if needed.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/node/startup.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/utils/injectabletime"
+	"github.com/aws/karpenter/pkg/utils/node"
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const StartupTimeout = 15 * time.Minute
```
Comment (RetWolf) [inline pkg/controllers/node/startup.go:31]: Should this be something configurable? I'm not sure Karpenter should always assume that nodes startup in under 15 minutes. This does seem like a sane default though.

File: pkg/controllers/node/controller.go [hunk 2]
```diff
@@ -42,7 +42,7 @@ const controllerName = "node"
 func NewController(kubeClient client.Client) *Controller {
 	return &Controller{
 		kubeClient: kubeClient,
-		liveness:   &Liveness{kubeClient: kubeClient},
+		startup:    &Startup{kubeClient: kubeClient},
```
Comment (ellistarn) [inline pkg/controllers/node/controller.go:45]: Thoughts on "initialization". I know it deviates a bit from `startupProbe`, but I think it's a more accurate term, and we're already not perfectly following the convention.

File: pkg/controllers/node/startup.go (hunk 1 above)
Comment (ellistarn) [reply pkg/controllers/node/startup.go:31]: We're trying to balance simplicity and configurability. One way to achieve this is to start with a good default, and expose configurability when it becomes necessary for a customer. Do you have a >15 min use case?

File: pkg/controllers/node/startup.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	for _, taint := range n.Spec.Taints {
+		if taint.Key != v1alpha5.NotReadyTaintKey {
+			taints = append(taints, taint)
+		}
+	}
+	n.Spec.Taints = taints
+	return reconcile.Result{}, nil
+}
+
+func notReadyTaintKeyExists(n *v1.Node) bool {
```
Comment (ellistarn) [inline pkg/controllers/node/startup.go:67]: Thoughts on making a similar helper to `taints.Has`. Perhaps `taints.HasKey(v1alpha5.NotReadyTaintKey)`

File: pkg/controllers/node/startup.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	if !notReadyTaintKeyExists(n) {
+		// At this point, the startup of the node is complete and no more evaluation is necessary.
+		return reconcile.Result{}, nil
+	}
+
+	if !node.IsReady(n) {
+		if timeSinceCreation := injectabletime.Now().Sub(n.GetCreationTimestamp().Time); timeSinceCreation < StartupTimeout {
+			return reconcile.Result{RequeueAfter: StartupTimeout - timeSinceCreation}, nil
+		}
+		logging.FromContext(ctx).Infof("Triggering termination for node that failed to transition to ready")
```
Comment (ellistarn) [inline pkg/controllers/node/startup.go:51]:
```suggestion
		logging.FromContext(ctx).Infof("Triggering termination for node that failed to become ready")
```

File: pkg/controllers/node/startup.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+// Reconcile reconciles the node
+func (r *Startup) Reconcile(ctx context.Context, _ *v1alpha5.Provisioner, n *v1.Node) (reconcile.Result, error) {
+	if !notReadyTaintKeyExists(n) {
+		// At this point, the startup of the node is complete and no more evaluation is necessary.
+		return reconcile.Result{}, nil
+	}
+
+	if !node.IsReady(n) {
+		if timeSinceCreation := injectabletime.Now().Sub(n.GetCreationTimestamp().Time); timeSinceCreation < StartupTimeout {
```
Comment (ellistarn) [inline pkg/controllers/node/startup.go:48]:
```suggestion
		if age := injectabletime.Now().Sub(n.GetCreationTimestamp().Time); age < StartupTimeout {
```

**Review by ellistarn — COMMENTED** [review]
Nice!

File: pkg/controllers/node/startup.go (hunk 1 above)
Comment (RetWolf) [reply pkg/controllers/node/startup.go:31]: No, I don't. That makes sense to me

File: pkg/controllers/node/controller.go (hunk 2 above)
Comment (suket22) [reply pkg/controllers/node/controller.go:45]: I don't mind. I'll change this to `initialization`

File: pkg/controllers/node/startup.go (hunk 3 above)
Comment (suket22) [reply pkg/controllers/node/startup.go]: Done, thanks for the suggestion!

File: pkg/controllers/node/initialization.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	if !v1alpha5.Taints(n.Spec.Taints).HasKey(v1alpha5.NotReadyTaintKey) {
+		// At this point, the startup of the node is complete and no more evaluation is necessary.
+		return reconcile.Result{}, nil
+	}
+
+	if !node.IsReady(n) {
+		if age := injectabletime.Now().Sub(n.GetCreationTimestamp().Time); age < InitializationTimeout {
+			return reconcile.Result{RequeueAfter: InitializationTimeout - age}, nil
+		}
+		logging.FromContext(ctx).Infof("Triggering termination for node that failed to become ready")
```
Comment (bwagner5) [inline pkg/controllers/node/initialization.go:51]: I think it would be helpful to log the node name here, wdyt?

Comment (ellistarn) [reply pkg/controllers/node/initialization.go:51]: This is included in our logger. All reconcilers include resource name in their logger at the top level. It helps avoid redundant statements.

**Review by ellistarn — APPROVED** [review]

**Pull request merged**